src/yunohost/log.py

```diff
@@ -0,0 +1,436 @@
+
+            # Start to record or give the unit operation in argument to let the
+            # developper start the record itself
+            if auto:
+                uo.start()
+            try:
+                if not auto:
+                    args = (uo,) + args
+                result = func(*args, **kwargs)
+            except Exception as e:
+                uo.error(e)
+                raise
+            else:
+                uo.success()
+            return result
+        return func_wrapper
+    return decorate
+
+class UnitOperation(object):
+    """
+    Instances of this class represents unit operation the yunohost admin as done.
+
+    Each time an action of the yunohost cli/api change the system, one or several
+    unit operations should be registered.
+
+    This class record logs and some metadata like context or start time/end time.
+    """
+
+    def __init__(self, operation, related_to=None, **kwargs):
+        # TODO add a way to not save password on app installation
+        self.operation = operation
+        self.related_to = related_to
+        self.extra = kwargs
+        self.started_at = None
+        self.ended_at = None
+        self.logger = None
+        self._name = None
+
+        self.path = OPERATIONS_PATH
+
+        if not os.path.exists(self.path):
+            os.makedirs(self.path)
+
+    def start(self):
+        """
+        Start to record logs that change the system
+        Until this start method is run, no unit operation will be registered.
+        """
+
+        if self.started_at is None:
+            self.started_at = datetime.now()
+            self.flush()
+            self._register_log()
+
+    def _register_log(self):
+        """
+        Register log with a handler connected on log system
+        """
+
+        # TODO add a way to not save password on app installation
+        filename = os.path.join(self.path, self.name + LOG_FILE_EXT)
+        self.file_handler = FileHandler(filename)
+        self.file_handler.formatter = Formatter('%(asctime)s: %(levelname)s - %(message)s')
+
+        # Listen to the root logger
+        self.logger = getLogger('yunohost')
+        self.logger.addHandler(self.file_handler)
+
+    def flush(self):
+        """
+        Write or rewrite the metadata file with all metadata known
+        """
+
+        filename = os.path.join(self.path, self.name + METADATA_FILE_EXT)
+        with open(filename, 'w') as outfile:
+            yaml.safe_dump(self.metadata, outfile, default_flow_style=False)
+
+    @property
+    def name(self):
+        """
+        Name of the operation
+        This name is used as filename, so don't use space
+        """
+        if self._name is not None:
+            return self._name
+
+        name = [self.started_at.strftime("%Y%m%d-%H%M%S")]
+        name += [self.operation]
+        if self.related_to:
+            if isinstance(self.related_to[0], tuple):
+                name += [self.related_to[0][1]]
```

Comment:
This behavior is kind of an issue imho. c.f. my other comment about adding a domain. I got the following message : 
```
Info: To view the log of the operation 'Regenerate system configurations 'nginx'', use the command 'yunohost log display 20180725-144847-service_regen_conf-nginx
```
so I had a look at the log and see a whole bunch of lines about metronome. So I was confused ... then I realize that nope, this isn't a regen of nginx. This is a regen of nginx+metronome+dnsmasq(+rmilter) ... 

This is a bit tricky do define what to do for this ... Naively I would do something like : "if there is only one related thing (e.g. service), use that, otherwise don't use anything". So in that case the description would just be "Regenerate configurations" and it's up to user / support to check the metadata. (Note that if you regen all config, some magic is used so that the description is "Regenerate configurations 'all'".)